pkg/kp/rcstore/store.go

```diff
@@ -31,7 +31,13 @@ type Store interface {
 	//
 	// The lock will be taken against an ephemeral key, so it is safe to use a
 	// session that deletes its keys on invalidation.
-	Lock(id fields.ID, session string) (bool, error)
+	//
+	// The separate locks are for one reader (who watches the RC and acts when
+	// it changes) and one writer (who mutates the RC to cause those changes).
+	// Each lock may be held by one, and only one, party (readers cannot share
+	// the read lock).
```

Comment:
I'm going to rewrite this section in the next PR, but it's still important to understand why so let me explain.

There are two entities interested in an RC: the entity that reads the RC and schedules/unschedules pods, and the entity that writes to the RC as part of some higher-layer abstraction. Obviously there can only be one writer. But the reader is also performing mutation on the layer of abstraction below the RC (preparers), and therefore there can only be one of those as well. If you had multiple goroutines running WatchDesires, you would get oscillation as each one tries to schedule over the other's changes. That's why there has to be mutual exclusion for the entity that reads the RC.